auth/api/iam/api.go

```diff
@@ -595,6 +608,203 @@ func (r Wrapper) StatusList(ctx context.Context, request StatusListRequestObject
+	authorizationDetails := []byte("[]")
+	if len(request.Body.AuthorizationDetails) > 0 {
+		authorizationDetails, _ = json.Marshal(request.Body.AuthorizationDetails)
+		if err != nil {
+			log.Logger().WithError(err).Errorf("failed to parse the authorization details")
+			return nil, err
+		}
+	}
+	// Generate the state and PKCE
+	state := uuid.NewString()
+	pkceParams := generatePKCEParams()
+	if err != nil {
+		log.Logger().WithError(err).Errorf("failed to create the PKCE parameters")
+		return nil, err
+	}
+	// Figure out our own redirect URL by parsing the did:web and extracting the host.
+	requesterDidUrl, err := didweb.DIDToURL(*requestHolder)
+	if err != nil {
+		log.Logger().WithError(err).Errorf("failed to create the PKCE parameters")
+		return nil, err
+	}
+	redirectUri, err := url.Parse("https://" + requesterDidUrl.Host + "/iam/oid4vci/callback")
+	if err != nil {
+		log.Logger().WithError(err).Errorf("failed to create the url for host: %s", requesterDidUrl.Host)
+		return nil, err
+	}
+	// Store the session
+	err = r.oid4vciSssionStore().Put(state, &Oid4vciSession{
+		HolderDid:   requestHolder.String(),
+		IssuerDid:   issuerDid.String(),
+		RedirectUrl: request.Body.RedirectUri,
+		RedirectUri: redirectUri.String(),
+		PKCEParams:  pkceParams,
+	})
+	if err != nil {
+		log.Logger().WithError(err).Errorf("failed to store the session")
+		return nil, err
+	}
+	// Build the redirect URL, the client browser should be redirected to.
+	redirectUrl := http2.AddQueryParams(*endpoint, map[string]string{
+		"response_type":         "code",
+		"state":                 state,
+		"client_id":             requestHolder.String(),
+		"authorization_details": string(authorizationDetails),
+		"redirect_uri":          redirectUri.String(),
+		"code_challenge":        pkceParams.Challenge,
+		"code_challenge_method": pkceParams.ChallengeMethod,
+	})
+
+	log.Logger().Debugf("generated the following redirect_uri for did %s, to issuer %s: %s", requestHolder.String(), issuerDid.String(), redirectUri.String())
+
+	return RequestOid4vciCredentialIssuance200JSONResponse{
+		RedirectUri: redirectUrl.String(),
+		SessionId:   state,
+	}, nil
+}
+
+func (r Wrapper) CallbackOid4vciCredentialIssuance(ctx context.Context, request CallbackOid4vciCredentialIssuanceRequestObject) (CallbackOid4vciCredentialIssuanceResponseObject, error) {
+	state := request.Params.State
+	oid4vciSession := Oid4vciSession{}
+	err := r.oid4vciSssionStore().Get(state, &oid4vciSession)
+	if err != nil {
+		return nil, err
+	}
+	if request.Params.Error != nil {
+		return r.errorResponse(oid4vciSession, *request.Params.Error, request.Params.ErrorDescription)
+	}
+	code := request.Params.Code
+	pkceParams := oid4vciSession.PKCEParams
+	issuerDid, err := did.ParseDID(oid4vciSession.IssuerDid)
+	if err != nil {
+		log.Logger().WithError(err).Errorf("could not resolve Issuer DID: %s", oid4vciSession.IssuerDid)
+		return r.errorResponse(oid4vciSession, "invalid_request", nil)
```

Comment:
we use `return nil, withCallbackURI(oauthError(oauth.ServerError, fmt.Sprintf("something something: %s", err.Error())), appCallbackURI)`

OAuthError types are picked up by the middleware and translated to redirects

Reply:
Done